src/docfx/build/Build.cs

```diff
@@ -47,19 +47,20 @@ private static List<Document> GlobFiles(Context context, Docset docset)
         {
             var manifest = new ConcurrentDictionary<Document, byte>();
             var references = new ConcurrentDictionary<Document, byte>();
+            var globbelFiles = new HashSet<Document>(files);
 
             await ParallelUtility.ForEach(
                 files,
                 (file, buildChild) =>
                 {
-                    if (!ShouldBuildFile(file, manifest, tocMap))
+                    if (!ShouldBuildFile(context, file, tocMap, globbelFiles) || !manifest.TryAdd(file, 0))
                     {
                         return Task.CompletedTask;
                     }
 
                     return BuildOneFile(context, file, tocMap, item =>
                     {
-                        if (ShouldBuildFile(item, references, tocMap))
+                        if (references.TryAdd(item, 0) && ShouldBuildFile(context, item, tocMap, globbelFiles))
```

Comment:
you don't need the additional `references` dictionary. It could be:

```
if (!ShouldBuildFile(file)) return
if (!manifest.TryAdd(file)) return

return BuildOneFile(item => if (ShouldBuildFile()) { xxx })
```

Where `ShouldBuildFile` has:

```
if (manifest.Contains(file)) return false
```

Reply:
there are a lot of files referenced by others more than 1 times, so the whole build pipeline will full of duplicated entities waiting to be queued, so we need filter them out from the beginning.

---
In reply to: [191297328](https://github.com/dotnet/docfx/pull/2818#discussion_r191297328) [](ancestors = 191297328)